contract/src/stage2/nft.rs

```diff
@@ -71,6 +71,48 @@ impl DropZone {
         PromiseOrValue::Value(false)
     }
 
+    #[private]
+    /// self callback checks if NFT was successfully transferred to the new account. If yes, do nothing. If no, refund original sender
+    pub fn nft_resolve_refund(
+        &mut self, 
+        drop_id: U128,
+        token_ids: Vec<String>, 
+    ) -> bool {
+        let used_gas = env::used_gas();
+        let prepaid_gas = env::prepaid_gas();
+
+        env::log_str(&format!("Beginning of resolve refund used gas: {:?} prepaid gas: {:?}", used_gas.0 / ONE_GIGGA_GAS, prepaid_gas.0 / ONE_GIGGA_GAS));
+
+        assert_eq!(env::promise_results_count(), 1, "no promise result");
+        let transfer_succeeded = matches!(env::promise_result(0), PromiseResult::Successful(_));
+        
+        // If not successful, the length of the token IDs needs to be added back to the drop.
+        if !transfer_succeeded {
```

Comment:
be careful here because it looks like the logic is reversed from the FT callback case

the token_ids are still present but the keys are not registered so they shouldn't be "claimable" now correct?

I don't think we have a double spend (or claim) of the NFTs though because you either would have:
- already sent the NFT back to the funder
- or the keys being unregistered would prevent a claim

Correct?

Reply:
The only case I'm worried about is the following: 
- NFT Drop with 50 keys (all registered) meaning there's 50 token IDs. 
- Assets refunded and so 50 keys are unregistered but the token IDs are still present. At this point, nobody can claim since there's 0 keys registered.
- In between the nft transfers and the callback, someone registers a key by sending an NFT.
- Someone claims a drop (still between the transfers and callback) which pops off the last token ID (that was already refunded but was still in the set). This will cause an unauthorized error and the NFT will not be transferred. Nothing will break on the contract but the user will not get an NFT.

Reply:
Is it much more expensive to remove all token IDs in the original call, and then add them all back in the callback?

Ideally, doesn't this come down to the drop owner, they should not be deleting keys and adding keys at the same time correct?

Future proofing, imagine a drop owner with many "associates" it theoretically, could happen.

For now wanna just throw in a TODO to add this to documentation so we can make owners aware and not lose sight of it?

Reply:
Yea i'll do that